types/underscore/index.d.ts

```diff
@@ -234,76 +239,139 @@ declare module _ {
-        * @param memo Initial reduce state.
-        * @param context `this` object in `iterator`, optional.
-        * @return Reduced object result.
-        **/
-        reduce<T, TResult>(
-            list: _.List<T>,
-            iterator: _.MemoIterator<T, TResult>,
-            memo?: TResult,
-            context?: any): TResult;
+         * Also known as inject and foldl, reduce boils down a collection of values into a
+         * single value. Memo is the initial state of the reduction, and each successive
+         * step of it should be returned by iteratee. The iteratee is passed four arguments:
+         * the memo, then the value and index (or key) of the iteration, and finally a reference
+         * to the entire collection.
+         *
+         * If no memo is passed to the initial invocation of reduce, the iteratee is not invoked
+         * on the first element of the collection. The first element is instead passed as the memo
+         * in the invocation of the iteratee on the next element in the collection.
+         * @param collection Reduces the elements of this collection.
+         * @param iteratee Reduce iteratee function for each element in `collection`.
+         * @param memo Initial reduce state or undefined to use the first collection item as initial state.
+         * @param context `this` object in `iteratee`, optional.
+         * @returns The reduced result.
+         **/
+        reduce<V extends Collection<any>>(
+            collection: V,
+            iteratee: MemoCollectionIterator<TypeOfCollection<V>, TypeOfCollection<V>, V>,
+            memo?: undefined,
+            context?: any
+        ): TypeOfCollection<V>;
+
+        /**
+         * Also known as inject and foldl, reduce boils down a collection of values into a
+         * single value. Memo is the initial state of the reduction, and each successive
+         * step of it should be returned by iteratee. The iteratee is passed four arguments:
+         * the memo, then the value and index (or key) of the iteration, and finally a reference
+         * to the entire collection.
+         *
+         * If no memo is passed to the initial invocation of reduce, the iteratee is not invoked
+         * on the first element of the collection. The first element is instead passed as the memo
+         * in the invocation of the iteratee on the next element in the collection.
+         * @param collection Reduces the elements of this collection.
+         * @param iteratee Reduce iteratee function for each element in `collection`.
+         * @param memo Initial reduce state or undefined to use the first collection item as initial state.
+         * @param context `this` object in `iteratee`, optional.
+         * @returns The reduced result.
+         **/
+        reduce<V extends List<any>, TResult>(
+            collection: V,
+            iteratee: MemoIterator<TypeOfCollection<V>, TResult, V>,
+            memo: TResult,
```

Comment:
I have two reasons for preferring an overload that specifically handles the undefined case here. First, an undefined memo isn't friendly to type inference when it provides context for `TResult` (see examples below). Second, it's nice to force the collection item type into the result in the case of an undefined memo to more accurately portray Underscore's behavior in the single-item collection case.

```TS
const result = _.reduce([1, 2, 3], (memo, num) => memo + num); // error - memo is of type "unknown"
const result = _.reduce([1, 2, 3], (memo, num) => memo + num, undefined, context); // error - memo is possibly "undefined" (which is its inferred type)
```

Reply:
Can't `TResult` be inferred from the iteratee instead of relying on the optional third argument?

Reply:
A `?` means that one can fail to provide the argument at all, and I don't think it's reasonable to allow that in this overload. [This test](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/45893#discussion_r449821966) could have done that and still gotten it wrong.

Reply:
To add a little more detail here, one can still provide `undefined` as the value for this argument, but the lack of `?` makes it so `undefined` has to be be provided explicitly rather than just leaving the argument out, which I think is appropriate for this overload.

Reply:
Despite the test that you linked to and the extra details, I'm not following you.

Reply:
Oh, sorry, I'll try to summarize quick:

You're correct, `TResult` can be inferred from the iteratee, but if `memo` is omitted then this overload would allow the iteratee to specify any `TResult` whether or not it includes the collection item type, which can be dangerous since the first collection item ends up being the memo. The test I referenced is an example of where that occurred - `number` was specified as `TResult`, giving `memo` the type `number`, but since the first item it was given was the first string in the collection and it treated it like a `number` when it should have treated it like a `string | number` the actual JS result was the string `'123'`.

Also, since the other overload handles the case where `memo` is not specified, this overload should require `memo` because that's what makes it different from the other overload.